reps/2023-06-30-serve-cpp-deployment.md

```diff
@@ -0,0 +1,411 @@
+    std::unordered_map<std::string, float> result;
+    for (auto it = recommend_cache_.begin(); it != recommend_cache_.end(); it++) {
+      result.insert({it->first, ComputeCosineSimilarity(request_vec, it->second)});
+    }
+    return result;
+  }
+
+  static SimilarityService *FactoryCreate() {
+    return new SimilarityService();
+  }
+
+ private:
+  float ComputeCosineSimilarity(std::vector<float> v1, std::vector<float> v2) {
+    int len = v1.size();
+    float dotProduct = 0;
+    float magnitude1 = 0;
+    float magnitude2 = 0;
+    for (int i = 0; i < len; i++) {
+      dotProduct += v1[i] * v2[i];
+      magnitude1 += v1[i] * v1[i];
+      magnitude2 += v2[i] * v2[i];
+    }
+    magnitude1 = std::sqrt(magnitude1);
+    magnitude2 = std::sqrt(magnitude2);
+    return dotProduct / (magnitude1 * magnitude2);
+  }
+
+  std::unordered_map<std::string, std::vector<float>> recommend_cache_ = {
+    {"mac", {1.5, 2.3, 3.5, 5.5}},
+    {"car", {1.5, 3.2, 3.9, 7.5}},
+    {"phone", {1.5, 2.0, 4.5, 8.1}},
+  };
+};
+
+// Register function
+SERVE_DEPLOYMENT(SimilarityService::FactoryCreate);
+
+}  // namespace serve
+}  // namespace ray
+
+```
+RankService:
+```cpp
+#pragma once
+
+#include <vector>
+#include <string>
+#include <queue>
+#include <unordered_map>
+#include <algorithm>
+
+namespace ray {
+namespace serve {
+
+class RankService {
+ public:
+  std::vector<std::string> Rank(std::unordered_map<std::string, float> recommends, int num) {
+    std::priority_queue<std::pair<std::string, float>, std::vector<std::pair<std::string, float>>, std::greater<std::pair<std::string, float>>> pq;
+    for (auto& pair : recommends) {
+        pq.push(pair);
+        if (pq.size() > num) {
+            pq.pop();
+        }
+    }
+
+    std::vector<std::string> result;
+    while (!pq.empty()) {
+        result.push_back(pq.top().first);
+        pq.pop();
+    }
+    std::reverse(result.begin(), result.end());
+    return result;
+  }
+
+  static RankService *FactoryCreate() {
+    return new RankService();
+  }
+};
+
+// Register function
+SERVE_DEPLOYMENT(RankService::FactoryCreate);
+
+}  // namespace serve
+}  // namespace ray
+
+```
+RecommendService is a sequential invocation of other services without complex processing logic, so we can directly use the DAG ability to connect these services, eliminating the need for RecommendService and simplifying user logic. 
+Next, we start the Ray Serve runtime and use Python Serve API deploy these Service as Deployment: 
+```python
+feature_service = serve.deployment(_func_or_class='FeatureService::FactoryCreate', name='feature_service', language='CPP')
+similarity_service = serve.deployment(_func_or_class='SimilarityService::FactoryCreate', name='similarity_service', language='CPP')
+rank_service = serve.deployment(_func_or_class='RankService::FactoryCreate', name='rank_service', language='CPP')
+
+with InputNode() as input:
```

Comment:
As discussed on the [Java improvement REP](https://github.com/ray-project/enhancements/pull/42#pullrequestreview-1607233173), let's please update this to use `bind(` with `DeploymentHandle` calls instead of the call graph API (there should be no `InputNode`)